### PR TITLE
Comments: Move changeCommentStatus into the data layer

### PR DESCRIPTION
**Pull request by Copons**
Move the `changeCommentStatus` implementation to the data layer, and replaces the `wpcom` API handler with the `http` middleware.
Remove `COMMENTS_CHANGE_STATUS_SUCCESS` and `COMMENTS_CHANGE_STATUS_FAILURE`, and makes full use of the `dispatchRequest` handlers.

The comment status is changed immediately after dispatching the `COMMENTS_CHANGE_STATUS` action.
On success, the comment status will be update again with the value returned by the API call.
On failure, the comment status will be updated back to its original value and an error notice will be displayed, with a "Try Again" button.

## Note

I'd like to use `apiNamespace` instead of `apiVersion`, but I can't seem to get it right.
I tried `wpcom/v1.1`, `wp/v1.1`, etc, but it's always invalid.

## Test

Check that changing a comment status still works as expected.

- Comment Management
- Post List
- Happychat? Not sure about this one, but it just seems a listener to the `_SUCCESS` action. cc @mattwondra

cc @Automattic/lannister

### CODE REVIEW

File: client/state/data-layer/wpcom/sites/comments/change-status/index.js [hunk 1]
```diff
@@ -27,24 +29,30 @@ const changeCommentStatus = ( { dispatch }, action ) => {
...
-	dispatch( {
-		...action,
-		status: get( data, 'status' ),
-	} );
+	dispatch(
+		local( {
+			...action,
+			status: get( data, 'status' ),
+		} )
+	);
```
Comment (dmsnell) [inline client/state/data-layer/wpcom/sites/comments/change-status/index.js]: we may want to dispatch this _only if_ the data in the response from the API call doesn't match what we find in state…

```js
const fromState = getCommentStatus( getState(), action.siteId, action.commentId );
const fromServer = data.comment_status;

if ( fromState !== fromServer && data.status ) {
	dispatch( local( {
		...action,
		status: data.status,
	} ) );
}
```

File: client/state/data-layer/wpcom/sites/comments/change-status/index.js [hunk 2]
```diff
@@ -0,0 +1,68 @@
...
+};
+
+const announceFailure = ( { dispatch, getState }, action ) => {
+	dispatch(
+		local( {
+			...action,
+			status: action.previousStatus,
+		} )
+	);
+	dispatch( errorNotice( translate( 'Could not update the comment' ) ) );
```
Comment (dmsnell) [inline client/state/data-layer/wpcom/sites/comments/change-status/index.js]: We could be quite a bit more helpful here, providing more context…

```
We could not approve the comment, "Actually I would argue that dogs are more suit…" Please [try again].
```

```
We were unable to trash the comment, "That's plain wrong; salamanders take the cake w…" [Try again?]
```

I _think_ we can include links here which could reissue the actions. The links are a nice future enhancement. The context is easy here to include.

File: client/state/data-layer/wpcom/sites/comments/change-status/index.js [hunk 3]
```diff
@@ -0,0 +1,98 @@
...
+				? translate( 'Could not unapprove the comment "%(commentContent)s".', { args: { commentContent } } )
+				: translate( 'Could not unapprove the comment.' );
+		case 'spam':
+			return commentContent
+				? translate( 'Could not mark the comment "%(commentContent)s" as spam.', { args: { commentContent } } )
+				: translate( 'Could not mark the comment as spam.' );
+		case 'trash':
+			return commentContent
+				? translate( 'Could not move the comment "%(commentContent)s" to trash.', { args: { commentContent } } )
+				: translate( 'Could not move the comment to trash.' );
```
Comment (a8ci18n) [inline client/state/data-layer/wpcom/sites/comments/change-status/index.js]: Hi! I've found a possible matching string that has already been translated **80** times:
`translate( 'Move this comment to the Trash' )` *ES Score: 8*
See 1 additional suggestion in the [PR translation status page](https://translate.wordpress.com/localci/status/automattic/wp-calypso/update/data-layer/change-comment-status)

Help me improve these suggestions: react with :thumbsdown: if the suggestion doesn't make any sense, or with :thumbsup: if it's a particularly good one (even if not implemented).

File: client/state/data-layer/wpcom/sites/comments/change-status/index.js [hunk 4]
```diff
@@ -0,0 +1,98 @@
...
+				? translate( 'Could not approve the comment "%(commentContent)s".', { args: { commentContent } } )
+				: translate( 'Could not approve the comment.' );
+		case 'unapproved':
+			return commentContent
+				? translate( 'Could not unapprove the comment "%(commentContent)s".', { args: { commentContent } } )
+				: translate( 'Could not unapprove the comment.' );
+		case 'spam':
+			return commentContent
+				? translate( 'Could not mark the comment "%(commentContent)s" as spam.', { args: { commentContent } } )
+				: translate( 'Could not mark the comment as spam.' );
```
Comment (a8ci18n) [inline client/state/data-layer/wpcom/sites/comments/change-status/index.js]: Hi! I've found a possible matching string that has already been translated **98** times:
`translate( 'Mark this comment as spam' )` *ES Score: 8*
See 1 additional suggestion in the [PR translation status page](https://translate.wordpress.com/localci/status/automattic/wp-calypso/update/data-layer/change-comment-status)

Help me improve these suggestions: react with :thumbsdown: if the suggestion doesn't make any sense, or with :thumbsup: if it's a particularly good one (even if not implemented).

File: client/state/data-layer/wpcom/sites/comments/change-status/index.js [hunk 5]
```diff
@@ -0,0 +1,98 @@
...
+
+	switch ( status ) {
+		case 'approved':
+			return commentContent
+				? translate( 'Could not approve the comment "%(commentContent)s".', { args: { commentContent } } )
+				: translate( 'Could not approve the comment.' );
+		case 'unapproved':
+			return commentContent
+				? translate( 'Could not unapprove the comment "%(commentContent)s".', { args: { commentContent } } )
+				: translate( 'Could not unapprove the comment.' );
```
Comment (a8ci18n) [inline client/state/data-layer/wpcom/sites/comments/change-status/index.js]: Hi! I've found a possible matching string that has already been translated **98** times:
`translate( 'Unapprove this comment' )` *ES Score: 12*
See 1 additional suggestion in the [PR translation status page](https://translate.wordpress.com/localci/status/automattic/wp-calypso/update/data-layer/change-comment-status)

Help me improve these suggestions: react with :thumbsdown: if the suggestion doesn't make any sense, or with :thumbsup: if it's a particularly good one (even if not implemented).

File: client/state/data-layer/wpcom/sites/comments/change-status/index.js [hunk 6]
```diff
@@ -0,0 +1,98 @@
...
+};
+
+const getErrorMessage = ( status, content ) => {
+	const commentContent = content && truncate( decodeEntities( stripHTML( content ) ), { length: 30, omission: '…' } );
+
+	switch ( status ) {
+		case 'approved':
+			return commentContent
+				? translate( 'Could not approve the comment "%(commentContent)s".', { args: { commentContent } } )
+				: translate( 'Could not approve the comment.' );
```
Comment (a8ci18n) [inline client/state/data-layer/wpcom/sites/comments/change-status/index.js]: Hi! I've found a possible matching string that has already been translated **31** times:
`translate( 'Approve this comment.' )` *ES Score: 8*
See 2 additional suggestions in the [PR translation status page](https://translate.wordpress.com/localci/status/automattic/wp-calypso/update/data-layer/change-comment-status)

Help me improve these suggestions: react with :thumbsdown: if the suggestion doesn't make any sense, or with :thumbsup: if it's a particularly good one (even if not implemented).

**Comment by Copons** [comment]
@dmsnell added more informative error message and the retry logic.

This is the list of new translation strings, to be checked and confirmed, I think, by both editorial (cc @Automattic/editorial) and i18n.

| Action | With Content | Without Content |
| --- | --- | --- |
| **Approve** | Could not approve the comment "%(commentContent)s". | Could not approve the comment. |
| **Unapprove** | Could not unapprove the comment "%(commentContent)s". | Could not unapprove the comment. |
| **Spam** | Could not mark the comment "%(commentContent)s" as spam. | Could not mark the comment as spam. |
| **Trash** | Could not move the comment "%(commentContent)s" to trash. | Could not move the comment to trash. |
| **Update** | Could not update the comment "%(commentContent)s". | Could not update the comment. |

**Comment by dmsnell** [comment]
@Copons I think that's a real benefit to our customers. The only real feedback I have on the messages is that I'm not sure if we should also try to get the post title in there, either in general of when the comment has no content. on that note, can a comment have no content? maybe `truncate( comment.content.trim() )` would be enough because I think that if a comment has no text it won't be accepted. worth checking…

**Comment by Copons** [comment]
@dmsnell 
> can a comment have no content

It sort of can, if for example the content is just an image, and we strip the HTML.
Though the check is there also in case, for whatever reason, the _comment itself_ fails (it doesn't exist? it hasn't been loaded yet? wrong ID?).

**Comment by michelleweber** [comment]
Hola from Editorial! Where do these messages appear? I'd love for them to be more normal-reading sentence, like "We couldn't move this comment to the trash" or "This comment couldn't be moved to the trash," but don't want to make suggestions without the context.

**Comment by Copons** [comment]
@mattwondra Thanks! 🙇 

@michelleweber Hi! They appear as error notices, positioned as you can see in the following screenshots, depending to the size of the screen, as a feedback for the "change comment status" action.

| Small | Medium | Big |
| --- | --- | --- |
| screen shot 2017-07-28 at 18 37 32 | screen shot 2017-07-28 at 18 37 46 | screen shot 2017-07-28 at 18 37 58 |

We're using the first 30 characters of the comment, as a way to easily identify it, as comments do not have titles or any other obvious mark except their own content (and maybe, secondarily, the author name).
The error messages come in 2 flavours, one if we can show the comment content, and a more generic one otherwise.
I've chosen 30 characters as a limit for the displayed content because I feel that it's long enough to be clear, but not too much to span multiple lines (in English, of course) at any screen width.

Should we want to increase the message length, though, I'd maybe reconsider showing the content (or showing less characters), as the notice doesn't really play well with long messages.

**Comment by drw158** [comment]
I don't think we should have the comment content in there. Just go with the generic messages.

I would think of it like email or text. If you just sent a text, something like this would make more sense:

"Text message failed to send"

not "Could not send text message "Hey are you going to the store later beca..."

**Comment by michelleweber** [comment]
Thanks, @Copons! In that case, I would suggest using the more normal/conversational phrasing "We couldn't move this comment to the trash" or "This comment couldn't be moved to the trash."

(And fwiw, I totally agree with @drw158 re: not including the comment content.)

**Comment by Copons** [comment]
Thanks for the suggestions!
I've updated the copy to these (I liked the active voice better):

- We couldn't approve this comment.
- We couldn't unapprove this comment.
- We couldn't mark this comment as spam.
- We couldn't move this comment to trash.
- We couldn't update this comment. (fallback if for some reason the action is not recognized)

File: client/state/data-layer/wpcom/sites/comments/change-status/index.js [hunk 7]
```diff
@@ -0,0 +1,85 @@
...
+			...action,
+			status: previousStatus,
+		} )
+	);
+
+	const errorMessage = {
+		approved: translate( "We couldn't approve this comment." ),
+		unapproved: translate( "We couldn't unapprove this comment." ),
+		spam: translate( "We couldn't mark this comment as spam." ),
+		trash: translate( "We couldn't move this comment to trash." ),
```
Comment (a8ci18n) [inline client/state/data-layer/wpcom/sites/comments/change-status/index.js]: Hi! I've found a possible matching string that has already been translated **80** times:
`translate( 'Move this comment to the Trash' )` *ES Score: 7*
See 1 additional suggestion in the [PR translation status page](https://translate.wordpress.com/localci/status/automattic/wp-calypso/update/data-layer/change-comment-status)

Help me improve these suggestions: react with :thumbsdown: if the suggestion doesn't make any sense, or with :thumbsup: if it's a particularly good one (even if not implemented).

File: client/state/data-layer/wpcom/sites/comments/change-status/index.js [hunk 8]
```diff
@@ -0,0 +1,85 @@
...
+		local( {
+			...action,
+			status: previousStatus,
+		} )
+	);
+
+	const errorMessage = {
+		approved: translate( "We couldn't approve this comment." ),
+		unapproved: translate( "We couldn't unapprove this comment." ),
+		spam: translate( "We couldn't mark this comment as spam." ),
```
Comment (a8ci18n) [inline client/state/data-layer/wpcom/sites/comments/change-status/index.js]: Hi! I've found a possible matching string that has already been translated **98** times:
`translate( 'Mark this comment as spam' )` *ES Score: 8*
See 1 additional suggestion in the [PR translation status page](https://translate.wordpress.com/localci/status/automattic/wp-calypso/update/data-layer/change-comment-status)

Help me improve these suggestions: react with :thumbsdown: if the suggestion doesn't make any sense, or with :thumbsup: if it's a particularly good one (even if not implemented).

File: client/state/data-layer/wpcom/sites/comments/change-status/index.js [hunk 9]
```diff
@@ -0,0 +1,85 @@
...
+	dispatch(
+		local( {
+			...action,
+			status: previousStatus,
+		} )
+	);
+
+	const errorMessage = {
+		approved: translate( "We couldn't approve this comment." ),
+		unapproved: translate( "We couldn't unapprove this comment." ),
```
Comment (a8ci18n) [inline client/state/data-layer/wpcom/sites/comments/change-status/index.js]: Hi! I've found a possible matching string that has already been translated **98** times:
`translate( 'Unapprove this comment' )` *ES Score: 13*
See 1 additional suggestion in the [PR translation status page](https://translate.wordpress.com/localci/status/automattic/wp-calypso/update/data-layer/change-comment-status)

Help me improve these suggestions: react with :thumbsdown: if the suggestion doesn't make any sense, or with :thumbsup: if it's a particularly good one (even if not implemented).

File: client/state/data-layer/wpcom/sites/comments/change-status/index.js [hunk 10]
```diff
@@ -0,0 +1,85 @@
...
+
+	dispatch(
+		local( {
+			...action,
+			status: previousStatus,
+		} )
+	);
+
+	const errorMessage = {
+		approved: translate( "We couldn't approve this comment." ),
```
Comment (a8ci18n) [inline client/state/data-layer/wpcom/sites/comments/change-status/index.js]: Hi! I've found a possible matching string that has already been translated **31** times:
`translate( 'Approve this comment.' )` *ES Score: 9*
See 1 additional suggestion in the [PR translation status page](https://translate.wordpress.com/localci/status/automattic/wp-calypso/update/data-layer/change-comment-status)

Help me improve these suggestions: react with :thumbsdown: if the suggestion doesn't make any sense, or with :thumbsup: if it's a particularly good one (even if not implemented).

File: client/state/data-layer/wpcom/sites/comments/change-status/index.js [hunk 11]
```diff
@@ -0,0 +1,85 @@
...
+				path: `/sites/${ siteId }/comments/${ commentId }`,
+				apiVersion: '1.1',
+				body: {
+					status,
+				},
+			},
+			{
+				...action,
+				previousStatus,
+			}
```
Comment (gwwar) [inline client/state/data-layer/wpcom/sites/comments/change-status/index.js]: Did we need the action here, it looks like we have a success/failure handler defined.

Comment (Copons) [inline client/state/data-layer/wpcom/sites/comments/change-status/index.js]: Do you mean that if I don't explicitly re-send the action in `http()`, it's just implicitly passed along to the success and failure handlers?

(Either way, I need to add the `previousStatus` to the action, in order to be able to set it back on failure.)

File: client/state/data-layer/wpcom/sites/comments/change-status/index.js [hunk 12]
```diff
@@ -0,0 +1,85 @@
...
+const changeCommentStatus = ( { dispatch, getState }, action ) => {
+	const { siteId, commentId, status } = action;
+	const previousStatus = get( getSiteComment( getState(), action.siteId, action.commentId ), 'status' );
+
+	dispatch(
+		http(
+			{
+				method: 'POST',
+				path: `/sites/${ siteId }/comments/${ commentId }`,
+				apiVersion: '1.1',
```
Comment (gwwar) [inline client/state/data-layer/wpcom/sites/comments/change-status/index.js]: What happens if you use apiNamespace?

**Comment by gwwar** [comment]
Testing the retry case I think we're missing a case:

To reproduce:
- Update the `http` call to an endpoint that doesn't exist (so it errors out)
- Unapprove a comment
- Wait for the notice to pop in
- Click on Retry
- I would expect status to be unapprove here:

screen shot 2017-07-31 at 1 01 30 pm

Basic functionality appears to work though 👍 . Feel free to 🚢 after we either remove the retry or fix this case.

File: client/state/data-layer/wpcom/sites/comments/change-status/index.js (hunk 11 above)
Comment (gwwar) [inline client/state/data-layer/wpcom/sites/comments/change-status/index.js]: I believe we should have access to the action regardless. If you look at `http` all it's doing is this. So if we didn't specify onSuccess for example, we'd auto dispatch this instead.
```jsx
		onSuccess: onSuccess || action,
		onFailure: onFailure || action,
		onProgress: onProgress || action,
```

Comment (gwwar) [inline client/state/data-layer/wpcom/sites/comments/change-status/index.js]: Looking at other usages I'm probably wrong here, but let me know what you find out, and we can update the documentation later.

Comment (gwwar) [inline client/state/data-layer/wpcom/sites/comments/change-status/index.js]: I guess we always need it, I'm not seeing the error event fire without it.

Comment (gwwar) [inline client/state/data-layer/wpcom/sites/comments/change-status/index.js]: Does that sound about right @dmsnell ?

Comment (dmsnell) [inline client/state/data-layer/wpcom/sites/comments/change-status/index.js]: I think the purpose here is to augment the original action with supplemental information

**Comment by Copons** [comment]
@gwwar It actually is working as expected.

- Comment starts as: `{ status: 'approved' }`

screen shot 2017-07-31 at 22 27 09

- Original action: `{ status: 'sdfakljsh' }`.

screen shot 2017-07-31 at 22 27 22

- Enter the `http`: `{ status: 'sdfakljsh', previousStatus: 'approved' }`
- Error!
- Enter the onFailure, **we assign previousStatus to status** and so the retry dispatch ends up `{ status: 'approved', previousStatus: 'approved' }`.

screen shot 2017-07-31 at 22 27 29

I know it's totally counter intuitive, but I couldn't think of a better way of handling this.
Maybe we could drop `previousStatus` from the retry dispatch so it's more clear?

**Comment by gwwar** [comment]
So here's what I'm seeing:

Initial Unapprove Click:
screen shot 2017-07-31 at 3 50 03 pm

Error Handling:
screen shot 2017-07-31 at 3 50 10 pm

Retry:
screen shot 2017-07-31 at 3 50 17 pm

The retry is the issue. I expect that to be `unapproved` for status.

**Comment by Copons** [comment]
@gwwar Finally found the issue: passing along `action` without giving much thought, also means passing along the entire `action.meta`, which in this case actually prevented the dispatch on the retry button.
Now you can clearly see a new `WPCOM_HTTP_REQUEST` being dispatched and the the `COMMENTS_CHANGE_STATUS` as well.

As an additional fix: I've also dropped the on success handler which screwed up the undo system in Comment Management with race conditions leading to complete un-sync between state and server.

**Pull request merged**